File: download_models.py
#!/usr/bin/env python3
"""Download and verify the model artifacts required by MimicMotion.

Run from the repository root on Linux:
    python download_models.py

Optional environment variables:
    HF_TOKEN       Hugging Face access token, if the SVD repository requires it.
    HF_ENDPOINT    Hugging Face endpoint (for example, https://hf-mirror.com).
"""

from __future__ import annotations

import os
import sys
import urllib.request
from pathlib import Path

try:
    from huggingface_hub import snapshot_download
except ImportError as exc:
    raise SystemExit(
        "Missing dependency. Install it with: "
        "python3 -m pip install 'huggingface_hub>=0.28,<1'"
    ) from exc

ROOT = Path(__file__).resolve().parent
MODELS_DIR = ROOT / "models"
SVD_DIR = MODELS_DIR / "SVD" / "stable-video-diffusion-img2vid-xt-1-1"

CORE_ARTIFACTS = (
    (
        "DWPose/yolox_l.onnx",
        "https://huggingface.co/yzd-v/DWPose/resolve/main/yolox_l.onnx?download=true",
        216_746_733,
    ),
    (
        "DWPose/dw-ll_ucoco_384.onnx",
        "https://huggingface.co/yzd-v/DWPose/resolve/main/dw-ll_ucoco_384.onnx?download=true",
        134_399_116,
    ),
    (
        "MimicMotion_1-1.pth",
        "https://huggingface.co/tencent/MimicMotion/resolve/main/MimicMotion_1-1.pth?download=true",
        3_049_867_447,
    ),
)

SVD_REQUIRED_FILES = (
    "model_index.json",
    "feature_extractor/preprocessor_config.json",
    "image_encoder/config.json",
    "image_encoder/model.fp16.safetensors",
    "scheduler/scheduler_config.json",
    "unet/config.json",
    "vae/config.json",
    "vae/diffusion_pytorch_model.fp16.safetensors",
)


def format_bytes(size: int) -> str:
    value = float(size)
    for unit in ("B", "KiB", "MiB", "GiB", "TiB"):
        if value < 1024 or unit == "TiB":
            return f"{value:.2f} {unit}"
        value /= 1024
    raise AssertionError("unreachable")


def download_with_resume(relative_path: str, url: str, expected_size: int) -> None:
    destination = MODELS_DIR / relative_path
    destination.parent.mkdir(parents=True, exist_ok=True)

    current_size = destination.stat().st_size if destination.exists() else 0
    if current_size == expected_size:
        print(f"SKIP {relative_path}: already complete ({format_bytes(current_size)})")
        return
    if current_size > expected_size:
        print(f"RESET {relative_path}: local file is larger than expected")
        destination.unlink()
        current_size = 0

    headers = {"User-Agent": "MimicMotion-artifact-downloader/1.0"}
    if current_size:
        headers["Range"] = f"bytes={current_size}-"
        mode = "ab"
        print(f"RESUME {relative_path} from {format_bytes(current_size)}")
    else:
        mode = "wb"
        print(f"DOWNLOAD {relative_path} ({format_bytes(expected_size)})")

    request = urllib.request.Request(url, headers=headers)
    try:
        response = urllib.request.urlopen(request, timeout=120)
    except Exception:
        if current_size:
            print(f"Resume failed for {relative_path}; retrying from the beginning")
            destination.unlink(missing_ok=True)
            request = urllib.request.Request(
                url,
                headers={"User-Agent": "MimicMotion-artifact-downloader/1.0"},
            )
            response = urllib.request.urlopen(request, timeout=120)
            mode = "wb"
        else:
            raise

    downloaded = current_size if mode == "ab" else 0
    next_report = downloaded + 128 * 1024 * 1024
    with response, destination.open(mode) as output:
        while True:
            chunk = response.read(8 * 1024 * 1024)
            if not chunk:
                break
            output.write(chunk)
            downloaded += len(chunk)
            if downloaded >= next_report:
                print(
                    f"  {relative_path}: {format_bytes(downloaded)} / "
                    f"{format_bytes(expected_size)}"
                )
                next_report = downloaded + 128 * 1024 * 1024

    actual_size = destination.stat().st_size
    if actual_size != expected_size:
        raise RuntimeError(
            f"Invalid size for {relative_path}: expected {expected_size}, got {actual_size}. "
            "Run this script again to resume."
        )
    print(f"OK {relative_path}: {format_bytes(actual_size)}")


def download_svd() -> None:
    print("DOWNLOAD SVD repository")
    snapshot_download(
        repo_id="stabilityai/stable-video-diffusion-img2vid-xt-1-1",
        local_dir=SVD_DIR,
        token=os.environ.get("HF_TOKEN") or None,
        max_workers=2,
    )


def verify_svd() -> None:
    missing = []
    for relative_path in SVD_REQUIRED_FILES:
        path = SVD_DIR / relative_path
        if not path.is_file() or path.stat().st_size == 0:
            missing.append(str(path.relative_to(ROOT)))
    if missing:
        raise RuntimeError("Missing required SVD files:\n  " + "\n  ".join(missing))
    print("OK required SVD files")


def main() -> int:
    MODELS_DIR.mkdir(parents=True, exist_ok=True)
    for relative_path, url, expected_size in CORE_ARTIFACTS:
        download_with_resume(relative_path, url, expected_size)
    download_svd()
    verify_svd()
    total_size = sum(path.stat().st_size for path in MODELS_DIR.rglob("*") if path.is_file())
    print(f"All model downloads completed: {format_bytes(total_size)} in {MODELS_DIR}")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        raise SystemExit("Download interrupted. Run the script again to resume.") from None
    except Exception as exc:
        raise SystemExit(f"Download failed: {exc}") from exc

File: download_models.js
#!/usr/bin/env node
/**
 * Download and verify the model artifacts required by MimicMotion.
 *
 * Run from the repository root on Linux with Node.js 20+:
 *   node download_models.js
 *
 * Optional environment variables:
 *   HF_TOKEN       Hugging Face access token.
 *   HF_ENDPOINT    Hugging Face endpoint, e.g. https://hf-mirror.com.
 */

"use strict";

const fs = require("node:fs");
const fsp = require("node:fs/promises");
const path = require("node:path");
const { Readable, Transform } = require("node:stream");
const { pipeline } = require("node:stream/promises");

const ROOT = __dirname;
const MODELS_DIR = path.join(ROOT, "models");
const SVD_DIR = path.join(
  MODELS_DIR,
  "SVD",
  "stable-video-diffusion-img2vid-xt-1-1",
);
const HF_ENDPOINT = (process.env.HF_ENDPOINT || "https://huggingface.co").replace(
  /\/$/,
  "",
);

const ARTIFACTS = [
  {
    relativePath: "DWPose/yolox_l.onnx",
    repo: "yzd-v/DWPose",
    repoPath: "yolox_l.onnx",
    expectedSize: 216_746_733,
  },
  {
    relativePath: "DWPose/dw-ll_ucoco_384.onnx",
    repo: "yzd-v/DWPose",
    repoPath: "dw-ll_ucoco_384.onnx",
    expectedSize: 134_399_116,
  },
  {
    relativePath: "MimicMotion_1-1.pth",
    repo: "tencent/MimicMotion",
    repoPath: "MimicMotion_1-1.pth",
    expectedSize: 3_049_867_447,
  },
  {
    relativePath: "SVD/stable-video-diffusion-img2vid-xt-1-1/model_index.json",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "model_index.json",
    expectedSize: 496,
  },
  {
    relativePath:
      "SVD/stable-video-diffusion-img2vid-xt-1-1/feature_extractor/preprocessor_config.json",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "feature_extractor/preprocessor_config.json",
    expectedSize: 518,
  },
  {
    relativePath:
      "SVD/stable-video-diffusion-img2vid-xt-1-1/image_encoder/config.json",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "image_encoder/config.json",
    expectedSize: 685,
  },
  {
    relativePath:
      "SVD/stable-video-diffusion-img2vid-xt-1-1/image_encoder/model.fp16.safetensors",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "image_encoder/model.fp16.safetensors",
    expectedSize: 1_264_217_240,
  },
  {
    relativePath:
      "SVD/stable-video-diffusion-img2vid-xt-1-1/scheduler/scheduler_config.json",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "scheduler/scheduler_config.json",
    expectedSize: 533,
  },
  {
    relativePath: "SVD/stable-video-diffusion-img2vid-xt-1-1/unet/config.json",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "unet/config.json",
    expectedSize: 984,
  },
  {
    relativePath: "SVD/stable-video-diffusion-img2vid-xt-1-1/vae/config.json",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "vae/config.json",
    expectedSize: 607,
  },
  {
    relativePath:
      "SVD/stable-video-diffusion-img2vid-xt-1-1/vae/diffusion_pytorch_model.fp16.safetensors",
    repo: "stabilityai/stable-video-diffusion-img2vid-xt-1-1",
    repoPath: "vae/diffusion_pytorch_model.fp16.safetensors",
    expectedSize: 195_531_910,
  },
];

function formatBytes(size) {
  let value = size;
  for (const unit of ["B", "KiB", "MiB", "GiB", "TiB"]) {
    if (value < 1024 || unit === "TiB") return `${value.toFixed(2)} ${unit}`;
    value /= 1024;
  }
  throw new Error("Unreachable");
}

function artifactUrl(artifact) {
  const encodedPath = artifact.repoPath
    .split("/")
    .map(encodeURIComponent)
    .join("/");
  return `${HF_ENDPOINT}/${artifact.repo}/resolve/main/${encodedPath}?download=true`;
}

function requestHeaders(offset = 0) {
  const headers = { "User-Agent": "MimicMotion-artifact-downloader/1.0" };
  const token = process.env.HF_TOKEN?.trim();
  if (token) headers.Authorization = `Bearer ${token}`;
  if (offset > 0) headers.Range = `bytes=${offset}-`;
  return headers;
}

class AuthenticationError extends Error {}

async function fetchWithRetries(url, offset, attempts = 10) {
  let lastError;
  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try {
      const response = await fetch(url, {
        headers: requestHeaders(offset),
        redirect: "follow",
        signal: AbortSignal.timeout(120_000),
      });
      if (response.status === 401 || response.status === 403) {
        throw new AuthenticationError(
          `Hugging Face returned HTTP ${response.status}. Accept the SVD model's ` +
            "access terms in your Hugging Face account, create a read token, and set HF_TOKEN.",
        );
      }
      if (response.status === 416) return response;
      if (!response.ok) {
        throw new Error(`HTTP ${response.status} ${response.statusText}`);
      }
      return response;
    } catch (error) {
      if (error instanceof AuthenticationError) throw error;
      lastError = error;
      if (attempt < attempts) {
        const delay = Math.min(attempt * 5_000, 30_000);
        console.warn(`  Request failed (${error.message}); retrying in ${delay / 1000}s`);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }
  }
  throw lastError;
}

async function downloadArtifact(artifact) {
  const destination = path.join(MODELS_DIR, artifact.relativePath);
  await fsp.mkdir(path.dirname(destination), { recursive: true });

  let currentSize = 0;
  try {
    currentSize = (await fsp.stat(destination)).size;
  } catch (error) {
    if (error.code !== "ENOENT") throw error;
  }

  if (currentSize === artifact.expectedSize) {
    console.log(
      `SKIP ${artifact.relativePath}: already complete (${formatBytes(currentSize)})`,
    );
    return;
  }
  if (currentSize > artifact.expectedSize) {
    console.log(`RESET ${artifact.relativePath}: file is larger than expected`);
    await fsp.rm(destination, { force: true });
    currentSize = 0;
  }

  console.log(
    `${currentSize ? "RESUME" : "DOWNLOAD"} ${artifact.relativePath}` +
      `${currentSize ? ` from ${formatBytes(currentSize)}` : ` (${formatBytes(artifact.expectedSize)})`}`,
  );

  let response = await fetchWithRetries(artifactUrl(artifact), currentSize);
  if (response.status === 416 && currentSize !== artifact.expectedSize) {
    await fsp.rm(destination, { force: true });
    currentSize = 0;
    response = await fetchWithRetries(artifactUrl(artifact), 0);
  }
  if (!response.body) throw new Error(`Empty response body for ${artifact.relativePath}`);

  let downloaded = currentSize;
  let nextReport = downloaded + 128 * 1024 * 1024;
  const progress = new Transform({
    transform(chunk, _encoding, callback) {
      downloaded += chunk.length;
      if (downloaded >= nextReport) {
        console.log(
          `  ${artifact.relativePath}: ${formatBytes(downloaded)} / ` +
            formatBytes(artifact.expectedSize),
        );
        nextReport = downloaded + 128 * 1024 * 1024;
      }
      callback(null, chunk);
    },
  });

  await pipeline(
    Readable.fromWeb(response.body),
    progress,
    fs.createWriteStream(destination, { flags: currentSize ? "a" : "w" }),
  );

  const actualSize = (await fsp.stat(destination)).size;
  if (actualSize !== artifact.expectedSize) {
    throw new Error(
      `Invalid size for ${artifact.relativePath}: expected ${artifact.expectedSize}, ` +
        `got ${actualSize}. Run this script again to resume.`,
    );
  }
  console.log(`OK ${artifact.relativePath}: ${formatBytes(actualSize)}`);
}

async function directorySize(directory) {
  let total = 0;
  for (const entry of await fsp.readdir(directory, { withFileTypes: true })) {
    const item = path.join(directory, entry.name);
    total += entry.isDirectory() ? await directorySize(item) : (await fsp.stat(item)).size;
  }
  return total;
}

async function main() {
  if (Number.parseInt(process.versions.node.split(".")[0], 10) < 20) {
    throw new Error("Node.js 20 or newer is required");
  }
  await fsp.mkdir(SVD_DIR, { recursive: true });
  for (const artifact of ARTIFACTS) await downloadArtifact(artifact);
  const total = await directorySize(MODELS_DIR);
  console.log(`All model downloads completed: ${formatBytes(total)} in ${MODELS_DIR}`);
}

main().catch((error) => {
  console.error(`Download failed: ${error.message}`);
  console.error("Run the script again to resume partial downloads.");
  process.exitCode = 1;
});
